src/D2L.CodeStyle.Analyzers/Immutability/ImmutableAttributeConsistencyChecker.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.CommonFixes;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	internal sealed class ImmutableAttributeConsistencyChecker {
+
+		internal delegate void DiagnosticSink( Diagnostic diagnostic );
+
+		private static readonly ImmutableDictionary<string, string> FixArgs = new Dictionary<string, string> {
+			{ AddAttributeCodeFix.USING_STATIC_ARG, "true" },
+			{ AddAttributeCodeFix.USING_NAMESPACE_ARG, "D2L.CodeStyle.Annotations.Objects" },
+			{ AddAttributeCodeFix.ATTRIBUTE_NAME_ARG, "Immutable" }
+		}.ToImmutableDictionary();
+
+		private readonly Compilation m_compilation;
+		private readonly DiagnosticSink m_diagnosticSink;
+		private readonly ImmutabilityContext m_context;
+
+		public ImmutableAttributeConsistencyChecker(
+			Compilation compilation,
+			DiagnosticSink diagnosticSink,
+			ImmutabilityContext context
+		) {
+			m_compilation = compilation;
+			m_diagnosticSink = diagnosticSink;
+			m_context = context;
+		}
+
+		public void CheckTypeDeclaration(
+			INamedTypeSymbol typeSymbol
+		) {
+			ImmutableTypeInfo typeInfo = m_context.GetImmutableTypeInfo( typeSymbol );
+
+			if( typeSymbol.BaseType != null ) {
+				CompareConsistencyToBaseType( typeInfo, m_context.GetImmutableTypeInfo( typeSymbol.BaseType ) );
+			}
+
+			foreach( INamedTypeSymbol interfaceType in typeSymbol.Interfaces ) {
+				CompareConsistencyToBaseType( typeInfo, m_context.GetImmutableTypeInfo( interfaceType ) );
+			}
+		}
+
+		public void CompareConsistencyToBaseType(
+			ImmutableTypeInfo typeInfo,
+			ImmutableTypeInfo baseTypeInfo
+		) {
+			switch( baseTypeInfo.Kind ) {
+				case ImmutableTypeKind.None:
+					return;
+
+				case ImmutableTypeKind.Instance:
+					return;
+
+				case ImmutableTypeKind.Total:
+					if( typeInfo.Kind == ImmutableTypeKind.Total ) {
+						break;
+					}

```

Comment:
Previous analyzer was checking for attributes. Pulling from context will catch implementers of the "extra types".

This will _probably_ blow up given IEnumerable and whatnot.

Reply:
Yep! Sure does!